### PR TITLE
Spinner: [a11y] support for status announcements

### LINKED ISSUE #1617: Spinners are silent to blind users
Environment: IE11 and JAWS 18
Are you willing to submit a PR to fix? No. I need help
Requested priority: High

Spinners should announce that progress is happening, but say nothing right now.

### PR DESCRIPTION
**Pull request by mdahamiwal**
#### Pull request checklist

- [x] Addresses an existing issue: Fixes #1617 
- [x] Include a change request file if publishing 
- [x] New feature, bugfix, or enhancement
  - [ ] Includes tests
- [x] Documentation update

#### Description of changes
Adding announcement capability to spinner used as progress indicators. If a label is provided the initial label and its updates will be fed to the screen reader.
Follows guidelines https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions

#### Focus areas to test

(optional)

### CODE REVIEW

**Comment by betrue-final-final** [comment]
Thank you Manish!

File: packages/office-ui-fabric-react/src/components/Spinner/Spinner.Props.ts [hunk 1]
```diff
@@ -25,14 +25,21 @@ export interface ISpinnerProps extends React.Props<Spinner> {
...
   /**
    * Additional CSS class(es) to apply to the Spinner.
    */
   className?: string;
+
+  /**
+   * Politeness setting for label update announcement.
+   * @default polite
+   */
+  ariaLive?: string;
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/Spinner/Spinner.Props.ts]: Should this be something like 'polite' | 'assertive' | etc.

**Comment by betrue-final-final** [comment]
I would expect polite

________________________________

File: packages/office-ui-fabric-react/src/components/Spinner/Spinner.Props.ts (hunk 1 above)
Comment (mdahamiwal) [inline packages/office-ui-fabric-react/src/components/Spinner/Spinner.Props.ts]: Yes, taking care of it.

**Comment by dzearing** [comment]
@betrue-final-final (BTW when you respond to these in outlook it gets totally garbled on github...)

I was referring to the typings. It should be a union of the possible values; rather than a string. There are only a very specific set of values that you should use for the aria-live attribute.

**Comment by dzearing** [comment]
@mdahamiwal should we be using the ScreenReaderAlert component in #676 here?

I noticed things it's doing and things you're missing here. For example, it uses `aria-atomic`. It also clears the text after 2 seconds. I don't know why.

Should we be committing #676 and then using that here? Is there value in having a ScreenReaderAlert component, or should we do what you've done here?

**Comment by mdahamiwal** [comment]
I have posted my comment on #676. It looks like a more complex and multi-purpose announcement control. I'm of opinion to expose it as a component to be used at places where we need fine-grain control over announcement like on a web page that doesn't use spinner or messagebar but they want announcement. I am not in favor of unnecessarily winding it with other components as that may lead to user confusion and usage issues.

**Comment by cschlechty** [comment]
Does it work better to just make the spinner an indeterminate progressbar? https://www.w3.org/TR/wai-aria/roles#progressbar

**Comment by mdahamiwal** [comment]
@cschlechty, it thought of doing it that way but it introduces another invisible component and we would need to provide special styles for that. Also, `aria-valuemin` and `aria-valuemax` doesn't make sense for a spinner. I wanted to keep the design simple without complicating the existing control. Although, if we see any major advantage with using progressbar I can consider. What you say?

**Comment by mdahamiwal** [comment]
@cschlechty, @dzearing , it would be great if we can close on this.

**Comment by cschlechty** [comment]
The progressbar role shouldn't need any extra hidden components. All spinners would just be indeterminate progress, which makes sense. By trying to do this via live-regions, you're going to quickly end up with region announcements clobbering each other.

**Comment by dzearing** [comment]
@cschlechty what would you propose then? Changing this PR to something else? Pushing it in? I'm not quite sure. I'd like to keep progress flowing.

**Comment by dzearing** [comment]
@mdahamiwal some questions:

1. What if you don't have a textual label, but you still want an aria live announcement?
2. What if there are multiple spinners on the page? Won't this devolve into an announcement mess (This is @cschlechty's point above.)

**Comment by dzearing** [comment]
An alternative approach would be that we do what you say, but also take in an `ariaLive` prop separately form the label. This would let the caller determine if this scenario is important for announcements.

**Comment by dzearing** [comment]
One of the things we were talking about is updating the Example Page stuff to have an accessibility section. Something like "Things the component provides" vs "Things you need to do to make your experience accessible". Guidelines for usage are super important so that we don't mislead customers into things "we're accessible by default because we use Fabric!"

**Comment by mdahamiwal** [comment]
Thanks everyone for really thinking this through and asking the right questions. I gave it a shot with an underlying `progressbar` and here is the differences between two approaches:

**Progressbar:** http://codepen.io/mdahamiwal/pen/oWowVd
- It doesn't announce the state change automatically. I tried it with NVDA, Narrator, I think the design pattern for progressbar governs that it only announces if progressbar is in focus.
- It prefixes the announcement with "progressbar" and percentages which defeats the purpose since our visual experience is spinner/loader, not progressbar. I guess progressbars are meant to convey quantitative progress in terms of percentages.

**Explicit live-region:** http://codepen.io/mdahamiwal/pen/YVEQEa
- Provides more control over announcements and politeness settings. We can turn it off if we don't want announcements.
- I have assigned a role of `status` which is more related to our spinner experience https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_status_role. I am open to feedback on this if you see any issue.
------------------------------------------------------------------------------------------------------------------
@dzearing 
> What if you don't have a textual label, but you still want an aria live announcement?

Good question. I guess we should add `ariaLabel` as property as well. This stands true even if label is different from announcement text. Contrarily, if you don't want announcement and pass a textual label, set `ariaLive="off"`.

> What if there are multiple spinners on the page? Won't this devolve into an announcement mess?

This is always the case and browser skips announcements. Eg. a progressbar updating status really fast, in that case only the last one gets announced or anything that has `ariaLive="assertive"` set. Same here, if you want to increase announcement priority for a particular spinner set its `ariaLive="assertive"`.

> An alternative approach would be that we do what you say, but also take in an ariaLive prop separately form the label. This would let the caller determine if this scenario is important for announcements.

I already have `ariaLive` exposed in the current changes to allow control over politeness settings.

**Proposal:** Keep current design with using ariaLive and DelayRendered and add option `ariaLabel` to support scenarios if there is no visible label or use need to override textual label. 

Thoughts ?

**Comment by betrue-final-final** [comment]
This may work for now, but I think we need to consider nonverbal audio supplements at some point, especially for indeterminate.

**Comment by betrue-final-final** [comment]
Correction, determinate

**Comment by mdahamiwal** [comment]
@cschlechty, can you provide your inputs? We need to close on this soon. Thanks

**Comment by cschlechty** [comment]
I'm okay with this change. If we get feedback that its too noisy, we can revisit.

**Comment by mdahamiwal** [comment]
@dzearing , @cschlechty, I have updated the PR, its ready for another look. Thanks

**Comment by jspurlin** [comment]
This really feel like a indeterminate `progressbar`  and a perfect example of a case where we should not be using the `aria-live` hammer to solve this issue. ATs should convey progress bar status to the user. We should follow up with the WPT and DEP to get their guidance on this (and get any bugs on their end logged)

**Comment by mdahamiwal** [comment]
@jspurlin, I understand your point but `progressbar` as aria pattern doesn't announce until the element is in focus, (check the deployed examples) and in this case we don't want spinners to be focusable. Secondly, it feels like a restricted pattern where it announces "progress bar" with every update which doesn't align with our spinner (loader) component.
This doesn't seem like a bug but inherent behavior so following up with WPT or DEP may not help but I am okay with initiating a thread to get more clarity on `progressbar` implementation. 

Coming to use of `aria-live`, I don't see it as a hammer or a complex pattern, in fact it's the basis of all other announcements (dialog, alertdialog, alert, status) including progressbar. So, I would say we are using more generic pattern which provides us more control on announcements settings.

**Comment by mdahamiwal** [comment]
@jspurlin , @dzearing, ping! Thanks

**Comment by jspurlin** [comment]
Since we don't want to set focus to the element, I'm fine with aria-live in this case

**Comment by mdahamiwal** [comment]
Thank you guys for your reviewing this.

**Pull request merged**